iroh-bytes/src/store/file.rs

```diff
@@ -0,0 +1,2483 @@
+    /// Outboard is in the inline_outboard table.
+    Inline(I),
+    /// Outboard is in the canonical location in the data directory.
+    Owned,
+    /// Outboard is not needed,
+    NotNeeded,
+}
+
+impl<I> OutboardLocation<I> {
+    fn discard_extra_data(self) -> OutboardLocation<()> {
+        match self {
+            Self::Inline(_) => OutboardLocation::Inline(()),
+            Self::Owned => OutboardLocation::Owned,
+            Self::NotNeeded => OutboardLocation::NotNeeded,
+        }
+    }
+}
+
+/// The information about an entry that we keep in the entry table for quick access.
+///
+/// The exact info to store here is TBD, so usually you should use the accessor methods.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub(crate) enum EntryState<I = ()> {
+    /// For a complete entry we always know the size. It does not make much sense
+    /// to write to a complete entry, so they are much easier to share.
+    Complete {
+        /// Location of the data.
+        data_location: DataLocation<I, u64>,
+        /// Location of the outboard.
+        outboard_location: OutboardLocation<I>,
+    },
+    /// Partial entries are entries for which we know the hash, but don't have
+    /// all the data. They are created when syncing from somewhere else by hash.
+    ///
+    /// As such they are always owned. There is also no inline storage for them.
+    /// Non short lived partial entries always live in the file system, and for
+    /// short lived ones we never create a database entry in the first place.
+    Partial {
+        /// Once we get the last chunk of a partial entry, we have validated
+        /// the size of the entry despite it still being incomplete.
+        ///
+        /// E.g. a giant file where we just requested the last chunk.
+        size: Option<u64>,
+    },
+}
+
+impl Default for EntryState {
+    fn default() -> Self {
+        Self::Partial { size: None }
+    }
+}
+
+impl EntryState {
+    fn union(self, that: Self) -> io::Result<Self> {
+        match (self, that) {
+            (
+                Self::Complete {
+                    data_location,
+                    outboard_location,
+                },
+                Self::Complete {
+                    data_location: b_data_location,
+                    ..
+                },
+            ) => Ok(Self::Complete {
+                // combine external paths if needed
+                data_location: data_location.union(b_data_location)?,
+                outboard_location,
+            }),
+            (a @ Self::Complete { .. }, Self::Partial { .. }) =>
+            // complete wins over partial
+            {
+                Ok(a)
+            }
+            (Self::Partial { .. }, b @ Self::Complete { .. }) =>
+            // complete wins over partial
+            {
+                Ok(b)
+            }
+            (Self::Partial { size: a_size }, Self::Partial { size: b_size }) =>
+            // keep known size from either entry
```

Comment:
why not prefer the larger one?

Reply:
Then you would be in so much trouble.

This is the validated size, which you will at some point set once you get the last chunk. This is currently not done at all, but I want to keep it in there so I don't have to perform a migration once I do it.

Reply:
both are validated? eg `a_size: Some(1024)` and `b_size: Some(2048)`, currently you choose the left one, so it depends on the calling order, is that the correct logic?

Reply:
Really the code should be

match (a_size, b_size) {
  Some(a), Some(b) if a != b { panic!("bao is broken"); }
}

Will add something along these lines.